### DISCUSSION TITLE
SMS OTP registration and login flow

### DISCUSSION THREAD

**Discussion by shrutikaCG** [Sunbird-RC/community#316] [General]
Hi Team,

We need help from sunbird team for the below concerns and queries:

1 - In the registration flow when user gives mobile number and generates the OTP, even before submitting and verifying the OTP the user gets created in Keycloak database. Whereas this works fine in the login flow.
 
2 - In both registration and login OTP page, whenever we refresh that page user gets new OTP that shouldn't be the case.
 
3 - We need to implement resend OTP link or button on the OTP page of both registration and login flow.
 
4 - To implement themes as per the business requirement, do we need to create new theme or can we use existing themes? If we need to create new theme can you please share the procedure or details for the same?

**Comment by tejash-jl** [comment]
> 1 - In the registration flow when user gives mobile number and generates the OTP, even before submitting and verifying the OTP the user gets created in Keycloak database. Whereas this works fine in the login flow.  

**you would need to write a custom layer to verify the mobile number before registration. Once the number is verified you can create the entity in registry**
>  2 - In both registration and login OTP page, whenever we refresh that page user gets new OTP that shouldn't be the case.   

**Since a OTP is already generated to the number and is stored in the cache for a short duration if the user refreshes the page, we need to trigger another OTP or else the timers will restart from the beginning.**
> 3 - We need to implement resend OTP link or button on the OTP page of both registration and login flow. 

**we already have resent otp button**  
> 4 - To implement themes as per the business requirement, do we need to create new theme or can we use existing themes? If we need to create new theme can you please share the procedure or details for the same?

**the theme source files are available here https://github.com/Sunbird-RC/sunbird-rc-core/tree/main/deps/keycloak/themes/sunbird-rc/login, You can check the below links on how to create custom login pages in keycloak https://www.baeldung.com/keycloak-custom-login-page https://www.opcito.com/blogs/customizing-login-and-signup-pages-in-keycloak**
